### PR TITLE
Feat | Store SDK

### PR DESCRIPTION
**Pull request by tlgimenes**
## What's the purpose of this pull request?
This PR adds the `store-sdk` package so we can share hooks easily between frameworks. Also, we move useMinicart hook to store-sdk and `once` function to make sure the tree shaking is working properly.

the Minicart hook was replaced by a global UI hook so we can add other global UI states in the same context, like openning a modal or not. A good read about it may be [this one](https://github.com/vercel/commerce/pull/199)

## How it works? 
<em>Tell us the role of the new feature, or component, in its context.</em>

## How to test it?
[btglobal](https://github.com/vtex-sites/btglobal.store/pull/488)
[marinbrasil](https://github.com/vtex-sites/marinbrasil.store/pull/484)
[storecomponents](https://github.com/vtex-sites/storecomponents.store/pull/834)
[carrefourbrfood](https://github.com/vtex-sites/carrefourbrfood.store/pull/524)

## References
<em>Spread the knowledge: is there any content you used to create this PR that is worth sharing?</em>  

<em>Extra tip: adding references to related issues or mentioning people important to this PR may be good for the documentation and reviewing process</em>

### CODE REVIEW

File: packages/gatsby-plugin-graphql/package.json [hunk 1]
```diff
@@ -34,5 +34,6 @@
     "@graphql-codegen/typescript-operations": "^1.17.7",
     "@graphql-tools/relay-operation-optimizer": "^6.0.15",
     "fs-extra": "^9.0.1"
-  }
+  },
+  "gitHead": "56d52a2f03d0041562b750ab8d021821f3ff2179"
```
Comment (klzns) [inline packages/gatsby-plugin-graphql/package.json:38]: Remove

File: packages/gatsby-source-vtex/package.json [hunk 2]
```diff
@@ -32,5 +32,6 @@
   "peerDependencies": {
     "gatsby": "^3.1.1",
     "graphql": "^14.0.0 || ^15.0.0"
-  }
+  },
+  "gitHead": "56d52a2f03d0041562b750ab8d021821f3ff2179"
```
Comment (klzns) [inline packages/gatsby-source-vtex/package.json:36]: Remove

File: packages/store-sdk/README.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
+# @vtex/store-sdk
+
+A simple, framework agnostic implementation of Commerce APIs to help you create you next React-based store with world class performance in record time
+
+[![NPM](https://img.shields.io/npm/v/@vtex/store-sdk.svg)](https://www.npmjs.com/package/@vtex/store-sdk) [![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)
+
+## Install
+
+```bash
+yarn add @vtex/store-ui
```
Comment (klzns) [inline packages/store-sdk/README.md:10]:
```suggestion
yarn add @vtex/store-sdk
```

File: packages/store-sdk/src/ui/Provider.tsx [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      throw new Error('Unknown UI state')
+    }
+  }
+}
+
+export const Provider: FC = ({ children }) => {
+  const [state, dispatch] = useReducer(reducer, initialState)
+
+  const openMinicart = () => dispatch({ type: 'OPEN_MINICART' })
+  const closeMinicart = () => dispatch({ type: 'CLOSE_MINICART' })
```
Comment (klzns) [inline packages/store-sdk/src/ui/Provider.tsx:54]: Might use useCallback so we can remove the `// eslint-disable-next-line react-hooks/exhaustive-deps`

**Review by klzns — CHANGES_REQUESTED** [review]

**Comment by tlgimenes** [comment]
@klzns , just added the comment.  Can you check it? Thanks!

**Comment by klzns** [comment]
> @klzns , just added the comment.  Can you check it? Thanks!

Quite impressed with the detail! Looks great

**Comment by tlgimenes** [comment]
@klzns, I was talking to @emersonlaurentino he was concerned on how the developer could extend the main UI context. To avoid using custom libs for state management, like recoil or redux, I tried implementing a new way of adding actions and effects to the UI's global state. 

Can you guys say what you think? I've also added some examples in the form of tests on how one could extend the global UI state

File: packages/store-sdk/README.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
+# @vtex/store-sdk
+
+A simple, framework agnostic implementation of Commerce APIs to help you create you next React-based store with world class performance in record time
```
Comment (emersonlaurentino) [inline packages/store-sdk/README.md:3]:
```suggestion
A simple, framework agnostic implementation of Commerce APIs to help you create your next React-based store with world class performance in record time
```

File: packages/store-sdk/package.json [hunk 6]
```diff
@@ -0,0 +1,47 @@
+{
+  "name": "@vtex/store-sdk",
+  "version": "0.328.0-alpha.1",
+  "description": "Hooks for creating your next component library",
```
Comment (emersonlaurentino) [inline packages/store-sdk/package.json:4]: component library?

File: packages/store-sdk/test/ui/Provider.test.tsx [hunk 7]
```diff
@@ -0,0 +1,135 @@
+import { act, renderHook } from '@testing-library/react-hooks'
+import React from 'react'
+import type { FC } from 'react'
+
+import { UIProvider, useGlobalUIState } from '../../src'
+import type { UIActions, UIEffects, UIInitializer } from '../../src'
+
+test('UI PRovider: Open/Close minicart', async () => {
```
Comment (emersonlaurentino) [inline packages/store-sdk/test/ui/Provider.test.tsx:8]:
```suggestion
test('UI Provider: Open/Close minicart', async () => {
```

File: packages/store-sdk/test/ui/Provider.test.tsx [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+  expect(result.current.displayMinicart).toBe(true)
+
+  act(() => {
+    result.current.closeMinicart()
+  })
+
+  expect(result.current.displayMinicart).toBe(false)
+})
+
+test('UI PRovider: Extend UI context', async () => {
```
Comment (emersonlaurentino) [inline packages/store-sdk/test/ui/Provider.test.tsx:26]:
```suggestion
test('UI Provider: Extend UI context', async () => {
```

File: packages/store-sdk/test/ui/Provider.test.tsx [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+  expect(result.current.displayModal).toBe(true)
+
+  act(() => {
+    result.current.closeModal()
+  })
+
+  expect(result.current.displayModal).toBe(false)
+})
+
+test('UI PRovider: Extend UI context with custom data', async () => {
```
Comment (emersonlaurentino) [inline packages/store-sdk/test/ui/Provider.test.tsx:67]:
```suggestion
test('UI Provider: Extend UI context with custom data', async () => {
```

**Review by emersonlaurentino — CHANGES_REQUESTED** [review]
What do you think about we rename `initializer` to `initialValues`? Initializer sounds to me like an action...

File: packages/store-sdk/package.json (hunk 6 above)
Comment (tlgimenes) [reply packages/store-sdk/package.json:4]: should be store, thanks

**Comment by tlgimenes** [comment]
> What do you think about we rename `initializer` to `initialValues`? Initializer sounds to me like an action...

I'm not sure. `InitialValues` seems like the second argument of `useReducer` hook. However this is the third. In react doc they use the `init` term. Do you prefer `init` here as well?

**Comment by tlgimenes** [comment]
Just changed initializer to init, like in react docs @emersonlaurentino :)

**Review by emersonlaurentino — APPROVED** [review]

**Pull request merged**